scripts/build/termux_step_massage.sh

```diff
@@ -60,27 +60,42 @@ termux_step_massage() {
 		fi
 	fi
 
-	local pattern=""
-	for file in ${TERMUX_PKG_NO_SHEBANG_FIX_FILES}; do
-		if [[ -z "${pattern}" ]]; then
-			pattern="${file}"
-			continue
+	if [ "$TERMUX_PKG_NO_SHEBANG_FIX" != "true" ]; then
+		local no_shebang_replace_full_regex="" no_shebang_replace_regex
+		# Assume files are a regex list and special characters in paths in each regex are already escaped.
+		if [[ -n "$TERMUX_PKG_NO_SHEBANG_FIX_FILES" ]]; then
+			while IFS= read -r no_shebang_replace_regex; do
+				[[ -n "$no_shebang_replace_full_regex" ]] && no_shebang_replace_full_regex+='|'
+				no_shebang_replace_full_regex+="($no_shebang_replace_regex)"
+			done < <(printf "%s\n" "$TERMUX_PKG_NO_SHEBANG_FIX_FILES")
+			if [[ -n "$no_shebang_replace_full_regex" ]]; then
+				no_shebang_replace_full_regex='(|./)('"$no_shebang_replace_full_regex"')$'
+			fi
 		fi
-		pattern+='|'"${file}"
-	done
-	if [[ -n "${pattern}" ]]; then
-		pattern='(|./)('"${pattern}"')$'
-	fi
 
-	if [ "$TERMUX_PKG_NO_SHEBANG_FIX" != "true" ]; then
+		local prefix_escaped shebang_already_valid_regex header_line shebang_match
+		local shebang_regex='^#!.*/bin/.*'
+
+		# Escape '\$[](){}|^.?+*' with backslashes.
+		prefix_escaped="$(printf "%s" "$TERMUX_PREFIX_CLASSICAL" | sed -zE -e 's/[][\.|$(){}?+*^]/\\&/g')"
+		shebang_already_valid_regex='^#! ?((/system/)|('"$prefix_escaped"'/))'
+
 		# Fix shebang paths:
-		while IFS= read -r -d '' file; do
-			if [[ -n "${pattern}" ]] && [[ -n "$(echo "${file}" | grep -E "${pattern}")" ]]; then
-				echo "INFO: Skip shebang fix for ${file}"
-				continue
-			fi
-			if head -c 100 "$file" | head -n 1 | grep -E "^#!.*/bin/.*" | grep -q -E -v -e "^#! ?/system" -e "^#! ?$TERMUX_PREFIX_CLASSICAL"; then
-				sed --follow-symlinks -i -E "1 s@^#\!(.*)/bin/(.*)@#\!$TERMUX_PREFIX/bin/\2@" "$file"
+		while IFS= read -rd '' file; do
+			read -rn$(( ${#TERMUX_PREFIX_CLASSICAL} + 4 )) header_line < "$file"
+			if [[ "${header_line:0:2}" == "#!" && "${#header_line}" -ge 3 && "$header_line" =~ $shebang_regex ]]; then
```

Comment:
"${#header_line}" -ge 3 is probably useless

Reply:
Instead of a regex match, a glob match should also be sufficient here.
```suggestion
			if [[ "$header_line" == "#!"*"/bin/"* ]]; then
```

Reply:
Length check exists to unnecessarily run a regex engine if at least a path doesn't exist after `#!`, but would be a rare occurrence.

Glob can't be used as regex is used to filter out with `$shebang_already_valid_regex`.

Reply:
This entire check is pretty messy.
Are you sure we can't simplify it with a glob in this case?

Reply:
Oh god...
No you're right we're using `${BASH_REMATCH[0]}` below.

Reply:
A glob could be used if prefix and /system checks are done with a glob too, but I don't have further time to rewrite and test or review everything again. And it shouldn't cause efficiency issues as would only run on files with a shebang.

Reply:
I think I'm ready to call it a mess and move on.